src/charm.py

```diff
@@ -5,98 +5,161 @@
-    "datahub-mae-consumer": "/bin/sh -c /datahub/datahub-mae-consumer/scripts/start.sh",
-    "datahub-mce-consumer": "/bin/sh -c /datahub/datahub-mce-consumer/scripts/start.sh",
-}
+APPLICATIONS: List[applications.AbstractApplication] = [
+    applications.ActionsApplication,
+    applications.FrontendApplication,
+    applications.GMSApplication,
+    applications.MAEConsumerApplication,
+    applications.MCEConsumerApplication,
+]
 
 
-CONTAINERS = list(COMMANDS.keys())
-
-
-def get_pebble_layer(application_name, context):
+def get_pebble_layer(
+    application: applications.AbstractApplication, context: applications.ApplicationContext
+) -> Dict:
     """Create pebble layer based on application.
 
     Args:
-        application_name: Name of DataHub application.
-        context: Environment to include with the pebble plan.
+        application: Name of DataHub application.
+        context: Context from which to gather environment to include with the pebble plan.
 
     Returns:
         pebble plan dict.
     """
     return {
         "summary": "datahub layer",
         "services": {
-            application_name: {
-                "summary": application_name,
-                "command": COMMANDS[application_name],
-                "startup": "enabled",
+            application.name: {
+                "summary": application.name,
+                "command": application.command,
+                "startup": "enabled" if application.is_enabled(context) else "disabled",
                 "override": "replace",
                 # Included to make configuration changes replan the service.
-                "environment": context,
+                "environment": application.compile_environment(context),
             },
         },
     }
 
 
-class DatahubK8SOperatorCharm(ops.CharmBase):
+class DatahubK8SOperatorCharm(TypedCharmBase[CharmConfig]):
     """Charm the application."""
 
+    config_type = CharmConfig
+
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
         self._state = State(self.app, lambda: self.model.get_relation("peer"))
 
-        for container in CONTAINERS:
-            self.framework.observe(self.on[container].pebble_ready, self._on_pebble_ready)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.peer_relation_changed, self._on_peer_relation_changed)
+        # TODO (mertalpt): Decide whether to implement or remove this.
+        # self.framework.observe(self.on.update_status, self._on_update_status)
+
+        # PostgreSQL
+        self.db = DatabaseRequires(
+            self, relation_name="db", database_name=DB_NAME, extra_user_roles="admin"
+        )
+        self.db_relation = PostgresqlRelation(self)
+
+        # Kafka
+        self.kafka = KafkaRequires(
+            self, relation_name="kafka", topic="datahub_dummy", extra_user_roles="admin"
```

Comment:
Any reason the topic is `datahub_dummy`? Is this not used? Also, this will not pass one of the CI tests in the future because of the use of `dummy`, better to use `placeholder` or other terms.